### PR TITLE
add `function.prototype.name`

### PR DESCRIPTION
**Pull request by ljharb**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Review by sheetalkamat — CHANGES_REQUESTED** [review]

File: types/function.prototype.name/implementation.d.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+type Func = (...args: any[]) => any;
+
+declare function name(): (this: Func) => string;
+
+export = name;
```
[the remark this answers is not in the document]
Comment (ljharb) [reply types/function.prototype.name/implementation.d.ts:5]: They are not implementation details; they're all explicit entry points that I use in dozens of packages, and are present in `exports`.

[the remark this answers is not in the document]
Comment (ljharb) [reply types/function.prototype.name/implementation.d.ts:5]: Rebased and repushed so the bot will do a proper rereview ping.

Do note that I have a hundred packages with this structure; some already are on DT; all of those have this DT file format already - so this is already an approved and accepted approach.

[the remark this answers is not in the document]
Comment (sheetalkamat) [reply types/function.prototype.name/implementation.d.ts:5]: https://github.com/es-shims/Function.prototype.name#example here talks about importing `function.prototype.name` so i dont understand why others are explicit entry points. It seems like these definitions could be simplified into index.d.ts

**Review by sheetalkamat — CHANGES_REQUESTED** [review]

File: types/function.prototype.name/implementation.d.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ljharb) [reply types/function.prototype.name/implementation.d.ts:5]: That's not how my entire ecosystem of shims works. The *example* only talks about that, but *all* of my shims have the same 4 entry points.

It feels like you're arguing about the design of my library, and not about the quality of the types. This is how my library is designed, this is how the types need to be for consumers to use it correctly. I'm not sure why this is up for debate?

[the remark this answers is not in the document]
Comment (ljharb) [reply types/function.prototype.name/implementation.d.ts:5]: Here's a few existing PRs where this exact (proper) pattern was followed:
 - #53498
 - #52716
 - #44406

[the remark this answers is not in the document]
Comment (ljharb) [reply types/function.prototype.name/implementation.d.ts:5]: Rebased this again to make sure it shows up in the right queue.

**Review by RyanCavanaugh — APPROVED** [review]

**Comment by ljharb** [comment]
Ready to merge

**Pull request merged**